enhancements/machine-api/control-plane-machinesets.md

```diff
@@ -0,0 +1,226 @@
+
+#### Story 7
+- As a developer, I want to be able to deploy the smallest possible cluster to save costs, i.e one instance Control Plane.
+
+### Implementation Details/Notes/Constraints [optional]
+
+To satisfy the goals, motivation and stories above this proposes to let the installer to create MachineSets to own the Machine objects for each failure domain and a Machine Health Checking to monitor the Control Plane machines.
+
+#### Bootstrapping
+Currently during a regular IPI bootstrapping process the installer uses Terraform to create a bootstrapping instance and 3 master instances. Then it creates Machine resources to "adopt" the existing master instances.
+
+Additionally:
+- It will create MachineSets to adopt those machines by looking up known labels (Adopting behaviour already exists in machineSet logic).
+	- `machine.openshift.io/cluster-api-MachineSet": <cluster_name>-<label>-<zone>-controlPlane`
+	- `machine.openshift.io/cluster-api-cluster":    clusterID`
+
+#### Declarative horizontal scaling
+- Each MachineSet can be scaled in/out by changing the replica number or via "scale" subresource.
+- Out of band the Cluster etcd Operator is responsible for watching the existing nodes and taking care of the etcd operational aspects.
+- During scaling in operations any machine deletion will always honour Pod Disruption Budgets (PBD). This gives etcd guard the chance to block a deletion that might be considered disruptive.
+
+#### Declarative Vertical scaling
+- With MachineSets in place this is semi-automated:
+	- A particular provider property can be changed by any consumer in the MachineSet spec.
+	- This won't affect existing Machines but it would apply for upcoming Machines.
+	- Machine creation with the new spec property can be forced by scaling out or deleting existing machines, e.g `kubectl delete machine name`
+
+#### Self healing (Machine Health Checking)
+- Specific MHC details can be found [here](https://github.com/openshift/enhancements/blob/master/enhancements/machine-api/machine-health-checking.md)
+
+- The Installer will create a Machine Health Checking resource targeting the Control Plane Machines.
+
+```
+apiVersion: machine.openshift.io/v1beta1
+kind: MachineHealthCheck
+metadata:
+  name: controlPlane 
+  namespace: openshift-machine-api
+spec:
+  selector:
+    matchLabels:
+      machine.openshift.io/cluster-api-machine-role: master
+      machine.openshift.io/cluster-api-machine-type: master
+  unhealthyConditions:
+  - type:    "Ready"
+    timeout: "300s"
+    status: "False"
+  - type:    "Ready"
+    timeout: "300s"
+    status: "Unknown"
+  maxUnhealthy: "34%"
+```
+
+-   Any machine deletion will always honour Pod Disruption Budgets (PBD). This gives etcd guard the chance to block a deletion that it considers disruptive.
+
+TODO: Currently multiple MHC can target the same Machine. We should revisit a stronger mechanisim to prevent undesired MHC to target Control Plane Machines.
```

Comment:
>Or ensure that no MHC maxUnhealthy is violated when determining if a remediation should occur 

That's current behaviour right? We want to prevent anything else setting a looser maxUnhealthy that includes master machines by accident.

Reply:
Current behaviour is to only check that the `maxUnhealthy` of the current MHC is satisfied, there's no check against other `maxUnhealthy`, so not, not current. My thinking was that when checking can a target be remediated, we could check the maxUnhealthy values for other MHCs that match this target, if any of those would block remediation, then remediation should be blocked. However, I think this would actually be pretty hard to implement because you'd have to health check all of the machines for each MHC to work out if their MaxUnhealthy would be violated.

We kinda need something more independent I think, maybe something like a MachineDisruptionBudget that could block deletions via a validating webhook similar to how a PDB works

Reply:
atm nothing prevents multiple MHC to monitor a single machine. For each of them the mhc controller will look at their particular maxUnhealthy, therefore the one with a looser value will remediate the machine. This TODO is to discuss how to prevent that from happening for masters. We don't want a MHC monitoring a master by accident to remediate the machine.
e.g >My thinking was that when checking can a target be remediated, we could check the maxUnhealthy values for other MHCs that match this target

Reply:
I don't think we have to over think this that much.  This is really a separate design, good to call it out, but don't block on it.

Reply:
Updated the proposal to have a managed MHC.